### PR TITLE
XML & XSD handling of Quorum

### PR DESCRIPTION
**Pull request by tombujok**
XML & XSD handling of Quorum.

Client-side (i.e. ClientDynamicClusterConfig) will be done in a separate PR. (https://github.com/hazelcast/hazelcast/pull/12127)

Also fixes **a lot** of omissions and bugs related to XML & XSD handling of data-structures (non quorum related).
  
Some of the unfixed ones have been documented as issues:

- #12117
- #12118
- #12119
- #12120 
- #12121

### CODE REVIEW

**Comment by tombujok** [comment]
run-lab-run

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastNamespaceHandler.java [hunk 1]
```diff
@@ -40,6 +40,10 @@ public void init() {
                 "set",
                 "list",
                 "executorService",
+                "durableExecutorService",
+                "scheduledExecutorService",
+                "ringBuffer",
+                "cardinalityEstimator",
```
Comment (tombujok) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastNamespaceHandler.java]: it's weird it worked without it...  I created an issue for the other missing ones: https://github.com/hazelcast/hazelcast/issues/12121

File: hazelcast-spring/src/main/resources/hazelcast-spring-3.10.xsd [hunk 2]
```diff
@@ -1774,6 +1791,18 @@
             </xs:appinfo>
         </xs:annotation>
     </xs:element>
+    <xs:element name="cardinalityEstimator" type="hazelcast-type">
+        <xs:annotation>
```
Comment (tombujok) [inline hazelcast-spring/src/main/resources/hazelcast-spring-3.10.xsd]: this was completely missing in...

File: hazelcast/src/test/java/com/hazelcast/config/ConfigCompatibilityChecker.java [hunk 3]
```diff
@@ -90,11 +90,13 @@ static boolean isCompatible(Config c1, Config c2) {
         checkCompatibleConfigs("map event journal", c1, c2, c1.getMapEventJournalConfigs(), c2.getMapEventJournalConfigs(), new MapEventJournalConfigChecker());
         checkCompatibleConfigs("cache event journal", c1, c2, c1.getCacheEventJournalConfigs(), c2.getCacheEventJournalConfigs(), new CacheEventJournalConfigChecker());
         checkCompatibleConfigs("multimap", c1, c2, c1.getMultiMapConfigs(), c2.getMultiMapConfigs(), new MultimapConfigChecker());
+        checkCompatibleConfigs("replicated map", c1, c2, c1.getReplicatedMapConfigs(), c2.getReplicatedMapConfigs(), new ReplicatedMapConfigChecker());
```
Comment (tombujok) [inline hazelcast/src/test/java/com/hazelcast/config/ConfigCompatibilityChecker.java]: This was completely missing in the check...

File: hazelcast/src/test/java/com/hazelcast/config/ConfigCompatibilityChecker.java [hunk 4]
```diff
@@ -90,11 +90,13 @@ static boolean isCompatible(Config c1, Config c2) {
...
         checkCompatibleConfigs("list", c1, c2, c1.getListConfigs(), c2.getListConfigs(), new ListConfigChecker());
         checkCompatibleConfigs("set", c1, c2, c1.getSetConfigs(), c2.getSetConfigs(), new SetConfigChecker());
         checkCompatibleConfigs("job tracker", c1, c2, c1.getJobTrackerConfigs(), c2.getJobTrackerConfigs(), new JobTrackerConfigChecker());
         checkCompatibleConfigs("reliable id generator", c1, c2, c1.getReliableIdGeneratorConfigs(), c2.getReliableIdGeneratorConfigs(), new ReliableIdGeneratorConfigChecker());
         checkCompatibleConfigs("count down latch", c1, c2, c1.getCountDownLatchConfigs(), c2.getCountDownLatchConfigs(), new CountDownLatchConfigChecker());
+        checkCompatibleConfigs("cardinality estimator", c1, c2, c1.getCardinalityEstimatorConfigs(), c2.getCardinalityEstimatorConfigs(), new CardinalityEstimatorConfigChecker());
```
Comment (tombujok) [inline hazelcast/src/test/java/com/hazelcast/config/ConfigCompatibilityChecker.java]: This was missing too...

File: hazelcast/src/test/java/com/hazelcast/config/ConfigXmlGeneratorTest.java [hunk 5]
```diff
@@ -196,14 +199,144 @@ public void testRingbuffer() {
         assertEquals(rbConfig, xmlRbConfig);
     }
 
+    @Test
```
Comment (tombujok) [inline hazelcast/src/test/java/com/hazelcast/config/ConfigXmlGeneratorTest.java]: Multiple tests were missing completely. I added them and created an issue to cover the missing attributes.
https://github.com/hazelcast/hazelcast/issues/12119

File: hazelcast/src/test/java/com/hazelcast/config/XMLConfigBuilderTest.java [hunk 6]
```diff
@@ -1403,6 +1411,39 @@ public void testQuorumListenerConfig() {
         assertEquals("com.abc.my.second.listener", quorumConfig.getListenerConfigs().get(1).getClassName());
     }
 
+    @Test
+    public void testCacheConfig() {
```
Comment (tombujok) [inline hazelcast/src/test/java/com/hazelcast/config/XMLConfigBuilderTest.java]: Here the same missing story. Documented as issue to add other attributes:
https://github.com/hazelcast/hazelcast/issues/12118

**Comment by tombujok** [comment]
run-lab-run

File: hazelcast/src/main/resources/hazelcast-full-example.xml [hunk 7]
```diff
@@ -573,11 +581,15 @@ https://hazelcast.org/documentation/.
         * <capacity>:
             Capacity of the scheduled executor. The default value is 100.
             This is the maximum number of tasks for a given scheduled executor on a particular partition.
+        * <quorum-ref>:
+            Adds the Split Brain Protection for this queue which you configure using the <quorum> element. You should set the <quorum-ref>'s value
```
Comment (ahmetmircik) [inline hazelcast/src/main/resources/hazelcast-full-example.xml]: this seems a copy&paste error all over the comments: `Adds the Split Brain Protection for this queue`, indeed this not a  queue?

Comment (tombujok) [reply hazelcast/src/main/resources/hazelcast-full-example.xml]: true, will fix that, thx!

File: hazelcast/src/main/java/com/hazelcast/config/ConfigXmlGenerator.java [hunk 8]
```diff
@@ -160,7 +160,8 @@ private static void collectionXmlGenerator(XmlGenerator gen, String type, Collec
                         .node("statistics-enabled", config.isStatisticsEnabled())
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/config/ConfigXmlGenerator.java]: Seems `mapConfigXmlGenerator` is missing the `quorum-ref`.

File: hazelcast/src/main/resources/hazelcast-config-3.10.xsd [hunk 9]
```diff
@@ -751,6 +751,7 @@
                     </xs:sequence>
                 </xs:complexType>
             </xs:element>
+            <xs:element name="quorum-ref" minOccurs="0" maxOccurs="1"/>
```
Comment (mmedenjak) [inline hazelcast/src/main/resources/hazelcast-config-3.10.xsd]: Can you add some documentation for all of these?

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java [hunk 10]
```diff
@@ -602,19 +605,51 @@ public void handleSymmetricEncryption(Node node, BeanDefinitionBuilder networkCo
...
+                if ("quorum-ref".equals(nodeName)) {
+                    execConfigBuilder.addPropertyValue("quorumName", getTextContent(childNode));
+                }
+            }
+            scheduledExecutorManagedMap.put(getAttribute(node, "name"), execConfigBuilder.getBeanDefinition());
         }
 
         public void handleCardinalityEstimator(Node node) {
-            createAndFillListedBean(node, CardinalityEstimatorConfig.class, "name", cardinalityEstimatorManagedMap);
+            BeanDefinitionBuilder execConfigBuilder = createBeanBuilder(CardinalityEstimatorConfig.class);
```
Comment (mmedenjak) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java]: The `createAndFillListedBean` also filled the `backup-count` and `async-backup-count` elements. Can you add those and add a test that verifies that parsing is still working as expected.

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastNamespaceHandler.java [hunk 11]
```diff
@@ -40,6 +40,10 @@ public void init() {
                 "set",
                 "list",
                 "executorService",
+                "durableExecutorService",
```
Comment (mmedenjak) [inline hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastNamespaceHandler.java]: `durableExecutorService` and `scheduledExecutorService` are not needed here or they should be added to `hazelcast-spring-3.10.xsd`. This list is for namespaces which allow the user to retrieve instances of distributed objects using just XML config, without calling methods on `HazelcastInstance`. 
For the current list, see:
https://github.com/hazelcast/hazelcast/blob/master/hazelcast-spring/src/main/resources/hazelcast-spring-3.10.xsd#L1608-L1836

This is different than a configuration for a distributed object which is created by the user later:
https://github.com/hazelcast/hazelcast/blob/master/hazelcast-spring/src/main/resources/hazelcast-spring-3.10.xsd#L41-L1440

So either add them to the XML section in the first link or remove them from here.

File: hazelcast-spring/src/main/resources/hazelcast-spring-3.10.xsd (hunk 2 above)
Comment (mmedenjak) [reply hazelcast-spring/src/main/resources/hazelcast-spring-3.10.xsd]: You can add the `durableExecutorService` and `scheduledExecutorService` as well.

Comment (tombujok) [reply hazelcast-spring/src/main/resources/hazelcast-spring-3.10.xsd]: Done

File: hazelcast/src/main/resources/hazelcast-full-example.xml (hunk 7 above)
Comment (tombujok) [reply hazelcast/src/main/resources/hazelcast-full-example.xml]: Fixed.

File: hazelcast/src/main/resources/hazelcast-config-3.10.xsd (hunk 9 above)
Comment (tombujok) [reply hazelcast/src/main/resources/hazelcast-config-3.10.xsd]: Done

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastNamespaceHandler.java (hunk 11 above)
Comment (tombujok) [reply hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastNamespaceHandler.java]: Added in XML and removed from here.

File: hazelcast/src/main/java/com/hazelcast/config/ConfigXmlGenerator.java (hunk 8 above)
Comment (tombujok) [reply hazelcast/src/main/java/com/hazelcast/config/ConfigXmlGenerator.java]: Lol, this is an old structure, so it was never there...

Comment (tombujok) [reply hazelcast/src/main/java/com/hazelcast/config/ConfigXmlGenerator.java]: Fixed.

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java (hunk 10 above)
Comment (tombujok) [reply hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastConfigBeanDefinitionParser.java]: That is true, that's a bug from my site. Fixed.

File: hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastNamespaceHandler.java (hunk 11 above)
Comment (mmedenjak) [reply hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastNamespaceHandler.java]: It needs to be in both XML and here or neither in XML nor here. Having it in just one of these two places doesn't work.

Comment (tombujok) [reply hazelcast-spring/src/main/java/com/hazelcast/spring/HazelcastNamespaceHandler.java]: Argh, added in both.

**Comment by tombujok** [comment]
@mmedenjak review comments addressed.

**Pull request merged**